File: Assets/Cryptor.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

using System;
using System.IO;
using System.Security.Cryptography;
using System.Text;

public class Cryptor
{
    public static string Encrypt(string plainText, string keyString, string dateNow)
    {
        byte[] cipherData;
        Aes aes = Aes.Create();
        aes.Key = Encoding.UTF8.GetBytes(keyString);
        //var randIv = new RNGCryptoServiceProvider();
        //randIv.GetBytes(new byte[16]);
        aes.GenerateIV(); //random iv generated
        Debug.Log("iv length " + aes.IV.Length);
        aes.Mode = CipherMode.CBC;
        ICryptoTransform cipher = aes.CreateEncryptor(aes.Key, aes.IV);

        using (MemoryStream ms = new MemoryStream())
        {
            using (CryptoStream cs = new CryptoStream(ms, cipher, CryptoStreamMode.Write))
            {
                using (StreamWriter sw = new StreamWriter(cs))
                {
                    sw.Write(plainText); // cipher update
                }
            }
            
            cipherData = ms.ToArray();
        }
        var utf8CipherData = Encoding.UTF8.GetString(cipherData, 0, cipherData.Length);
        Debug.Log("CipherData " + utf8CipherData);
        byte[] combinedData = new byte[aes.IV.Length + cipherData.Length]; //make the array for combined data, with the length of 
        Array.Copy(aes.IV, 0, combinedData, 0, aes.IV.Length); //masukin iv bytes nya ke array tadi
        Array.Copy(cipherData, 0, combinedData, aes.IV.Length, cipherData.Length); //lalu masukin cypherdata setelah iv nya, harusnya gak masukin ivnya?
        var encrypted = Convert.ToBase64String(combinedData); //encryption datanya harus sama
        var base64IV = Convert.ToBase64String(aes.IV);
        Debug.Log("date now in c# " + dateNow);
        Debug.Log("encrypted " + encrypted);
        Debug.Log("base64IV " + base64IV);
        var encryption = Convert.ToBase64String(Encoding.UTF8.GetBytes($"{utf8CipherData }|{dateNow}|{base64IV}"));
        Debug.Log("encription " + encryption);
        return encryption;
        //return Convert.ToBase64String(combinedData);
    }

    public static string Decrypt(string combinedString, string keyString)
    {
        string plainText;
        byte[] combinedData = Convert.FromBase64String(combinedString);
        Aes aes = Aes.Create();
        aes.Key = Encoding.UTF8.GetBytes(keyString);
        byte[] iv = new byte[aes.BlockSize / 8];
        byte[] cipherText = new byte[combinedData.Length - iv.Length];
        Array.Copy(combinedData, iv, iv.Length);
        Array.Copy(combinedData, iv.Length, cipherText, 0, cipherText.Length);
        aes.IV = iv;
        aes.Mode = CipherMode.CBC;
        ICryptoTransform decipher = aes.CreateDecryptor(aes.Key, aes.IV);

        using (MemoryStream ms = new MemoryStream(cipherText))
        {
            using (CryptoStream cs = new CryptoStream(ms, decipher, CryptoStreamMode.Read))
            {
                using (StreamReader sr = new StreamReader(cs))
                {
                    plainText = sr.ReadToEnd();
                }
            }

            return plainText;
        }
    }
}

File: Assets/cryptor.js
require('dotenv').config()

const { Console } = require('console');
const crypto = require('crypto')

class Cryptor {
  key

  constructor(key) {
    this.key = key;
  }

  encrypt(data, iv) {
    if (data == null) {
      throw new Error('Data is required')
    }

    try {
      // make the encrypter function
      const encrypter = crypto.createCipheriv("aes-256-cbc", this.key, iv);
      const base64IV = Buffer.from(iv).toString('base64')
      console.log("the iv is" , iv)
      // encrypt the message
      // set the input encoding
      // and the output encoding
      let encryptedWord = encrypter.update(data, "utf8", "hex");

      // stop the encryption using
      // the final method and set
      // output encoding to hex
      encryptedWord += encrypter.final("hex");
      console.log("encrycrypted word ",encryptedWord)
      console.log("date now ", Date.now())
      console.log("base64iv", base64IV)

      var encryption = Buffer.from(`${encryptedWord}|${Date.now()}|${base64IV}`).toString('base64')
      console.log("encripted is ", encryption);
      return encryption
    } catch (e) {
      throw new Error(e)
    }
  }

  decrypt(data) {
    try {
      // make the decrypter function
      const decrypter = crypto.createDecipheriv("aes-256-cbc", this.key, this.iv);

      // decrypt the message
      // set the input encoding
      // and the output encoding
      let decryptedWord = decrypter.update(data, "hex", "utf8");

      // stop the decryption using
      // the final method and set
      // output encoding to utf8
      decryptedWord += decrypter.final("utf8")

      return decryptedWord
    } catch (e) {
      throw new Error(e);
    }
  }
}

module.exports = Cryptor
